lib/shoryuken/queue.rb

```diff
@@ -83,7 +111,15 @@ def validate_message!(options)
       elsif !body.is_a?(String)
         fail ArgumentError, "The message body must be a String and you passed a #{body.class}"
       end
-
+      if is_fifo? && options[:delay_seconds].is_a?(Fixnum)
+        fail ArgumentError, 'FIFO queues do not accept DelaySeconds arguments.'
+      end
+      if is_fifo? && options[:message_group_id].nil?
+        fail ArgumentError, 'This queue is FIFO and no message_group_id was provided.'
+      end
+      if is_fifo? && !has_content_deduplication? && options[:message_deduplication_id].nil?
+        fail ArgumentError, 'This queue is FIFO without ContentBasedDeduplication enabled, and no MessageDeduplicationId was supplied'
+      end
       options
```

Comment:
@richseviora WDYT:

```ruby
    def validate_message!(options)
      body = options[:message_body]
      if body.is_a?(Hash)
        options[:message_body] = JSON.dump(body)
      elsif !body.is_a?(String)
        fail ArgumentError, "The message body must be a String and you passed a #{body.class}"
      end

      validate_fifo_message!(options)

      options
    end

    def validate_fifo_message!(options)
      return unless fifo?

      if options[:delay_seconds].is_a?(Fixnum)
        fail ArgumentError, 'FIFO queues do not accept DelaySeconds arguments.'
      end

      if options[:message_group_id].nil?
        fail ArgumentError, 'This queue is FIFO and no message_group_id was provided.'
      end

      if options[:message_deduplication_id].nil? && !has_content_deduplication?
        fail ArgumentError, 'This queue is FIFO without ContentBasedDeduplication enabled, and no MessageDeduplicationId was supplied'
      end
    end
```

This `if options[:message_deduplication_id].nil? && !has_content_deduplication?` condition change is because `has_content_deduplication?` can perform a request, so it's better to check `options[:message_deduplication_id]` first, because we may not need to check the other which's more expensive.